Comments, continued, strings, card as enum, etc

chrsoo opened the pull request:
This is a big change that may not be aligned with the vision for FITSRS. If so, no worries if you reject it.

Note that some sample files likely have invalid cards as per FITSv4, these cards are stored as `Card::Undefined`.

All unit tests pass, once the external sample file is downloaded and  installed.

# Changes
* Added support for continued strings (long-string).
* Added comments to `Value` enum.
* Refactored `Card` as enum with variants for the different card variants in FITSv4
* Changed `cards` in `Header` from `HashMap` to `Vec` to represent all cards in a FITS header.
* Added `values` `HashMap` in `Header` to hold `Cards` representing values
(i.e. Cards with with `= ` in bytes 9 and 10 of the buffer)
* Added `history()` and `comments()` for retrieving all HISTORY and COMMENT cards in the FITS header.
* Added `HISTORY` and `COMMENT` as keywords in `values` (this is redundant with the two previous methods
* Added support for value unit convention in comments
* ...

# Todo
* support for complex numbers (integer and float)
* support for HIEARCH cards
* additional tests and documentation
* some code cleanup
* ...

# Open Potins
* Having the HISTORY and COMMENT supported both by methods returning an iterator and as concatenated values in the keyword/value map is redundant and potentially confusing, there should probably only be one way to get the header HISTORY and COMMENT. Also COMMENT might not make sense as it is usually contextual and related to the adjoint value cards.
* There is potentially additional refactoring to be done for the extensions, e.g. switching to strings instead of eight byte buffers.

bmatthieu3 commented on src/card.rs:
```diff
@@ -1,22 +1,366 @@
+        comment: Option<String>,
+    },
+    // FITS extension string value, cf FITSv4, section 4.2.1.1
+    Extension([u8; 8]),
```
Should put an enum here Extension that could have only 3 possibilities, Image, Table and BinaryTable ?

chrsoo replied:
Yes, especially since there is already an enum defined in `fitsrs::hdu::header::extension`:
```rust
pub enum XtensionType {
    Image,
    BinTable,
    AsciiTable,
}
```

chrsoo replied:
Done.

bmatthieu3 commented on src/hdu/header/mod.rs:
```diff
@@ -250,26 +226,152 @@ where
+    /// [cards](Card) using the `HISTORY` keyword.
+    ///
+    pub fn history(&self) -> impl Iterator<Item = &String> + use<'_, X> {
+        self.filter_string_values_by_kw("HISTORY")
```
I may be wrong but it seems the filter_string_values_by_kw method only check for cards being the Value variant and do not consider the History variant (as well as COMMENT ...). So here the history cards might not be returned ?

bmatthieu3 replied:
Sorry I understand the comments and history are copied 2 times in the Card vec, one time as a history (resp comment)  variant and one time as a value variant. What was the reason for this redundancy ? I think that if possible, it would be nice indeed to store them only one time.

Maybe for history simply as a history variant. And for comment maybe the same. I am not very familiar whether a COMMENT card can refer to a value card or is more general ? I would say if someone wants to comment a specific card then he writes it in after the value after the / character ? (which is a case you handle too, associating a string comment as a value)

chrsoo replied:
> I may be wrong but it seems the filter_string_values_by_kw method only check for cards being the Value variant and do not consider the History variant (as well as COMMENT ...). So here the history cards might not be returned ?

You are very much right, not sure what I was thinking. Shows the value of unit tests and the peril of leaving them out. In this case the unit tests were on the TODO list, and I actually found it yesterday when I started making them. Also found a few other inconsistencies, like the fact that the SIMLPE and NAXIS headers (etc) are not present in the `values` map.

> Sorry I understand the comments and history are copied 2 times in the Card vec, one time as a history (resp comment) variant and one time as a value variant. What was the reason for this redundancy ? I think that if possible, it would be nice indeed to store them only one time.

My reasoning is that there should be a memory representation of the header as found in the FITS file which is why all header lines are (should be) in the `cards` vec. This would be useful later on for writing the header to disk, as it means it is possible to closely emulate the original content. Here is where there are two design decisions to make: a) is it a goal to preserve the header when writing, and b) is it preferable to do in-place editing of the file instead of writing back the whole header.

I do realise I might be thinking ahead a bit too much as FITS file writing is still in the future.

Given the above I see the `values` map as a convenience to the API user as the FITS long-strings values are already processed. We could of course prioritise memory over cpu cycles and process the long-strings on-demand, but I was thinking that we want to catch issues when parsing and not at retrieval of the values by API clients. The memory vs cpu I think is probably a non-issue.

Regarding the fact that HISTORY and COMMENT are stored in both `cards` and `values`, this is what I mentioned in the Open Points above. I did that just for the sake of the discussion, but when I started on the header unit tests I am thinking that adding them as concatenated strings (separated by `\n`) is actually a mistake. I think the on-demand filtering (now fixed and soon pushed) is better.

Retrieving HISTORY as a vector of strings make a lot of sense, since it is a processing log that may be displayed in a UI, but COMMENT I am less sure of as I understand this to be header comments in the file that may be interspersed between the values. They make sense when seen in the header, side-by-side with the values, but aggregated in a vec or as a concatenated string... the context is lost.

chrsoo replied:
* Removed HISTORY and COMMENT from the `values`
* Added missing unit tests
* Fixed some bugs
* Improved parsing
* Added support for comments on Xtension.

chrsoo replied:
So, if you prefer not to have an in-memory representation of the header as seen on disk (admittedly, e.g. five Cards representing space might not be that useful), I suggest that we as an alternative...

1. Remove `comments()` on the header and consider header-level comments not part of the API
2. Replace the `cards` field with `history` and parse all `history` cards upfront.

This means that anybody interested in the header-level comments would have to use another tool (e.g. a text editor).

If you think that it should be possible to see the header level comments using the API, I propose we leave the PR more or less as is, with the only change that we remove the `comments()` method as I believe it makes little sense to put all comments together. Better to display them inline with the rest of the unprocessed cards.

bmatthieu3 replied:
Hi @chrsoo, thank you very much for all the precision. Give me some time understanding/testing the code again, it it a big and very valuable PR :) . I think that is good you removed the HISTORY and COMMENT from values to remove the redundancy.

When will come the time to writing I think that is a good to indeed dump all the cards with little to no modification from the original read header. Maybe one modification I see would be to dump the mandatory cards in the good order given by the standard for each extension type.

I like that you give a special enum variant for HISTORY and COMMENT. I think I share your ideas, retrieving all the HISTORY cards might be useful but the COMMENTS, as they might relate to different things is strange.

bmatthieu3 commented:
I think I will fix/reorganize a little bit the code so that we indeed have the simple and bitpix cards inside the cards vec as well.
What I propose:
* we first fill the cards vec
* then process the cards to get the hashmap of values
* and finally create the xtension object to check. This process checks for mandatory values in case we have an image extension, or bintable etc...

chrsoo commented:
Sounds good to me - let me know if you want me to do anything. Happy to help.

BTW we might consider putting the foot down whether the API should use`&str`, `String`, or the custom `Keyword` (`[u8;8]`) type for the keyword names in the API function and method signatures.

My preference is to use `&str` for *API users*, but  byte arrays might still make more sense for internal use by `fitsrs` developers, at least in some cases like matching in a card buffer.

Changing `get` on the `Header` struct is a breaking change, but depending on how you approach versioning and where you consider the API to be today, might still be acceptable. I added `value` as an alternative to `get` in order not to break more things than necesary.

bmatthieu3 commented:
I agree with you, it is better to let the user just give a '&str'.

This project is not at any major version so I would say it is possible to break the API as it is a new project. I do not know if it is anywhere used appart from my use case inside (aladin lite)[https://github.com/cds-astro/aladin-lite] so I do not know if we would bother someone by breaking it.

Nonetheless having both these methods does not break the api and it adds the possibility for the user to give a `&str`. To not take any risk and as I want to release a version quite soon we could just say in the doc of the get method that it is deprecated (so keeping it in the api) but tell the user to use `value` instead and that get will eventually be removed from the api in favor of `value`

bmatthieu3 commented:
Also I did a commit and tried to push to the PR but I could not. May it be possible that I need some authorization or something like that ? It is also possible that I did something wrong so I will try tomorrow at work.

chrsoo commented:
> What do you think ?

If there is not too much concern about breaking changes, I would change `get` to use `&str` and delete `value` when releasing the next minor version. This breaking change should be OK even for a minor version, as `0.y.z` releases are [considered unstable](https://semver.org/#spec-item-4) in semantic versioning. Once you release `1.0.0` we should be more strict and increase the major version for such a change.

> Also I did a commit and tried to push to the PR but I could not. May it be possible that I need some authorization or something like that ? It is also possible that I did something wrong so I will try tomorrow at work.

I invited you as a collaborator on my personal `fitsrs` repository, but I am not sure how it is supposed to work. Best solution might be to create a branch in the upstream repository and work against that.

bmatthieu3 commented:
I pushed to your branch. Let me know if you do not agree with something otherwise I think we can merge (test are passing)

chrsoo commented:
> I pushed to your branch. Let me know if you do not agree with something otherwise I think we can merge (test are passing)

* Fixed an error where the SIMPLE keyword was ignored and added a unit test. 
* Improved documentation and deleted the non-sensical `comments` method on `Header`.

Managed to self-inflict a merge conflict, sorry for messing up the commit history...

Yes, I think it is ready to merge! :-)

bmatthieu3 commented:
Nice you fixed the error on SIMPLE. Let's merge it, that is a very good contribution thank you!
